app/services/bolognese/writers/hyku_addons_work_form_fields_writer.rb

```diff
@@ -3,29 +3,92 @@
 # NOTE:
 # Add new fields here that should be available to prefill the work forms.
 # You must also add the logic to the PrefillWorkFormViaDOI JS class.
+#
+# Raw output from this class can be accessed via - Copy the raw source, not the HTML output:
+# http://YOUR_TENANT.lvh.me:3000/doi/autofill?curation_concern=generic_work&doi=YOUR_DOI
+#
+# E.g
+# http://repo.lvh.me:3000/doi/autofill?curation_concern=generic_work&doi=10.7554/elife.63646
 module Bolognese
   module Writers
     module HykuAddonsWorkFormFieldsWriter
       DATE_FORMAT = "%Y-%-m-%-d"
+      DOI_REGEX = /10.\d{4,9}\/[-._;()\/:A-Z0-9]+/
+      ROR_QUERY_URL = "https://api.ror.org/organizations?query="
 
       def hyku_addons_work_form_fields
         {
-          'identifier' => Array(identifiers).select { |id| id["identifierType"] != "DOI" }.pluck("identifier"),
-          'doi' => Array(doi),
-          'title' => titles&.pluck("title"),
-          'creator' => write_involved("creators"),
-          'contributor' => write_involved("contributors"),
-          'publisher' => Array(publisher),
-          'date_created' => write_date("date_created", collect_date("Issued")),
-          'date_updated' => write_date("date_updated", collect_date("Updated")),
+          "identifier" => Array(identifiers).select { |id| id["identifierType"] != "DOI" }.pluck("identifier"),
+          "doi" => Array(doi),
+          "title" => titles&.pluck("title"),
+          "creator" => write_involved("creators"),
+          "contributor" => write_involved("contributors"),
+          "funder" => write_funders,
+          "publisher" => Array(publisher),
+          "date_created" => write_date("date_created", collect_date("Issued")),
+          "date_updated" => write_date("date_updated", collect_date("Updated")),
           "date_published" => write_date_published,
-          'description' => write_descriptions,
-          'keyword' => subjects&.pluck("subject")
+          "abstract" => write_descriptions,
+          "keyword" => subjects&.pluck("subject")
         }
       end
 
       protected
 
+        def write_funders
+          funding_references.map do |funder|
```

Comment:
I'm a bit concerned with composing these two together (DOI metadata fetching and ROR metadata fetching).  There could easily be another data source we want to integrate later and I don't think it makes sense to inline all of them.  I guess I was thinking at the least having the ROR api handling done in a separate service.  It would be good to figure out an interface for this.  Maybe there should be a new endpoint which has a new backing service class which orchestrates fetching all of this linked data in the right order and returning JSON for populating the deposit form.  It would be nice if it could also be used by backend processes as well like for the autopopulation in AH WP1.

Reply:
These are all fair points. I had been focusing on getting all this working, then could start to abstract it out.